File: solutions/python/space-age/2/space_age.py
from functools import partial

ORBITAL_PERIOD = {
    "mercury": 0.2408467,
    "venus": 0.61519726,
    "earth": 1.0,
    "mars": 1.8808158,
    "jupiter": 11.862615,
    "saturn": 29.447498,
    "uranus": 84.016846,
    "neptune": 164.79132,
}

SECONDS_PER_EARTH_YEAR = 31_557_600


class SpaceAge:
    _years: float

    def __init__(self, seconds: int) -> None:
        self._years = seconds / SECONDS_PER_EARTH_YEAR
        for planet in ORBITAL_PERIOD:
            setattr(self, "on_" + planet, partial(self._on_planet, planet))

    def _on_planet(self, planet: str) -> float:
        return round(self._years / ORBITAL_PERIOD[planet], 2)

File: solutions/c/space-age/2/space_age.c
#include "space_age.h"

#define EARTH_SECONDS(period) (period * 60 * 60 * 24 * 365.25)

const float ORBITAL_PERIOD[] = {
    EARTH_SECONDS(0.2408467),  // Mercury
    EARTH_SECONDS(0.61519726), // Venus
    EARTH_SECONDS(1.0),        // Earth
    EARTH_SECONDS(1.8808158),  // Mars
    EARTH_SECONDS(11.862615),  // Jupiter
    EARTH_SECONDS(29.447498),  // Saturn
    EARTH_SECONDS(84.016846),  // Uranus
    EARTH_SECONDS(164.79132),  // Neptune
};

float age(planet_t planet, int64_t seconds)
{
    if (planet < MERCURY || planet > NEPTUNE)
        return ERROR;

    return seconds / ORBITAL_PERIOD[planet];
}
